### ISSUE TITLE
V0-A — characterize Vulkan and native inference on AMD and NVIDIA

### ISSUE THREAD

**Issue by ezutfen** [Zutfen-LLC/inferswarm#141]
## Parent

#140 — V0 portable accelerator execution substrate evaluation

## Status

**AGENT-READY — PHYSICAL CHARACTERIZATION / RESEARCH ONLY**

This issue may begin immediately and in parallel with the active Issue #117 correctness/remediation line.

Authored against InferSwarm `main@d1afad64ca8bdd05634d3d0e1e09b8077da525d5`. At execution, record the actual starting `origin/main`; do not require main to remain at this SHA.

## Objective

Produce the first retained, matched physical evidence for Vulkan as an inference execution substrate on the current InferSwarm hardware.

Required hardware classes:

1. AMD GPUs installed in `inferswarm02`;
2. one available NVIDIA GPU on an InferSwarm node as a control.

Primary comparisons, where a credible native backend is actually available:

```text
same AMD GPU:    Vulkan  <-> HIP/ROCm
same NVIDIA GPU: Vulkan  <-> CUDA
```

Cross-vendor AMD-vs-NVIDIA numbers are descriptive only. The primary backend-cost comparisons are **same physical device, same software/model/workload, different backend**.

The purpose is not to make llama.cpp the InferSwarm runtime. It is a convenient external probe for current Vulkan/native backend correctness, materialization, memory, and workload economics before InferSwarm itself is changed.

## Architectural boundary

Do not change the InferSwarm planner, Fabric Doctrine, accepted runtime implementation, FreeToken research branch, or public architecture APIs in this issue.

Treat Vulkan/CUDA/HIP as observed execution capabilities of physical Compute Units, not as the Compute Unit identity itself.

Do not modify accepted #117/#133/#137 evidence or participate in Arm-C/Arm-D execution.

## Phase 0 — physical inventory before changes

On `inferswarm02`, before installing/upgrading GPU drivers or changing kernel/runtime state, retain the current physical/software inventory.

At minimum record:

- hostname and timestamp;
- OS/distribution/release;
- kernel;
- CPU and system RAM;
- every installed display/compute GPU from PCI enumeration;
- vendor/device/subsystem IDs;
- PCI BDF;
- negotiated and maximum PCIe width/speed where observable;
- NUMA relationship where applicable;
- BAR / resizable-BAR state where observable;
- current kernel driver for every GPU;
- Mesa version and relevant ICD packages;
- Vulkan loader version;
- installed Vulkan ICDs and exact ICD selected per device;
- `vulkaninfo` device identity and API version;
- Vulkan memory heaps/types and reported device-local capacity;
- subgroup size/properties;
- FP16 / integer-dot / cooperative-matrix or equivalent advertised capabilities relevant to inference;
- current ROCm/HIP presence and whether the installed GPU is recognized;
- temperatures/power limits only as environmental/provenance data, not performance claims.

Discover the exact AMD hardware; do not rely on issue text or maintainer recollection for SKU/VRAM.

If multiple AMD GPUs are installed:

- inventory every card;
- if all are materially identical, run the full campaign on one explicitly selected physical GPU and at least a smoke/repeatability check on the others;
- if they are materially different SKUs/VRAM/architecture revisions, characterize each distinct class far enough to determine whether conclusions differ.

On the NVIDIA control node, capture the corresponding identity/topology/driver/Vulkan/CUDA facts for the exact selected GPU.

Choose an NVIDIA device/node that can be exercised without disturbing an active correctness-bearing InferSwarm physical campaign. Record why the selected control was safe to use.

## Host-change discipline

The initial inventory is immutable evidence of the starting state.

Normal research tooling/build dependencies and Vulkan utilities may be installed as needed on `inferswarm02`. Record every material package/runtime change.

If Vulkan cannot run under the starting driver stack:

1. retain the failure evidence first;
2. identify the smallest credible driver/userspace correction;
3. record the change as an explicit intervention;
4. re-inventory after the change.

Do not silently rewrite the baseline.

Do not perform a major NVIDIA control-node driver/kernel upgrade merely for this experiment. If the existing NVIDIA stack cannot provide Vulkan, retain `CONTROL_BACKEND_BLOCKED` evidence and report it rather than destabilizing a known-good CUDA node.

## Phase 1 — freeze the external probe

Use a pinned `llama.cpp` commit as the first inference probe unless execution evidence demonstrates a concrete blocker requiring another tool.

Before performance/correctness collection, freeze and retain:

- exact llama.cpp repository + commit SHA;
- compiler/CMake/toolchain identity;
- build configuration;
- enabled backends;
- Vulkan shader/build options that materially affect execution;
- CUDA/HIP build options for native comparison;
- produced binary hashes;
- backend/device enumeration output.

Prefer the same llama.cpp commit across every backend/device.

Use `llama-bench` and `llama-cli` as the canonical first probes. `llama-server` may be characterized separately but must not replace CLI/bench results because server-layer regressions/overhead would confound the backend question.

Do not use llama.cpp multi-GPU performance as the primary InferSwarm verdict. Canonical V0-A performance runs are one physical GPU at a time unless a narrowly scoped diagnostic requires otherwise.

## Phase 2 — freeze a matched model/workload subject

Select one practical GGUF model/quantization that:

- is supported by the pinned build on both vendor paths being compared;
- fits completely enough on the smallest selected GPU to avoid host-offload becoming the dominant uncontrolled variable;
- is large enough to produce meaningful sustained prefill and decode work;
- does not require changing model bytes or quantization between backends.

Reuse an already-present suitable model if available; otherwise acquire one through a reproducible source.

Freeze before canonical runs:

- model repository/source;
- exact revision if applicable;
- exact GGUF filename;
- file size;
- SHA-256;
- architecture;
- parameter scale;
- quantization;
- context/batch settings;
- prompt/workload fixtures;
- generation settings/seeds where applicable.

If no single representation can truthfully run on every required device, define the largest common matched subject first. Additional per-device subjects may be diagnostic only and must not be used for same-model cross-backend ratios.

## Phase 3 — backend smoke and correctness

For each required `(physical GPU, backend)` pair:

1. prove the exact requested device/backend is selected rather than a silent CPU/other-GPU fallback;
2. prove substantial model state is resident on the intended GPU;
3. run deterministic prompt/generation fixtures repeatedly;
4. capture exact generated tokens;
5. capture logits or another practical numerical observable where the probe exposes them without materially changing execution;
6. capture NaN/Inf/failure state;
7. capture crashes, shader compile failures, unsupported-op fallbacks, backend warnings, and retries;
8. distinguish deterministic backend-local behavior from cross-backend numerical differences.

Do not require bit equality across CUDA/HIP/Vulkan unless the evidence shows it. Classify numerical behavior using the same conceptual separation InferSwarm uses elsewhere:

- exact model/input/integrity identity;
- numerical execution difference;
- semantic/generated-output behavior.

This issue does not set a new permanent heterogeneous numerical threshold. Retain enough raw evidence for V0-B to judge whether deeper qualification is warranted.

## Phase 4 — matched performance campaign

For every viable backend pair on the same physical GPU, run a matched campaign with held-constant model and workload parameters.

At minimum separate:

### Startup/materialization

- process startup;
- model load/materialization wall time;
- time to first ready state;
- peak/final GPU memory;
- peak/final host RSS;
- persistent host memory after model materialization;
- evidence of model-state duplication/staging where observable.

### Prefill / prompt processing

- prompt tokens;
- prompt-processing wall time;
- prompt tokens/s;
- GPU utilization/power where reliable telemetry exists;
- relevant memory occupancy.

### Decode / generation

- generated tokens;
- decode wall time;
- tokens/s;
- inter-token latency if directly available;
- GPU utilization/power where reliable;
- memory occupancy.

### Stability

- repeat count;
- run-to-run distribution;
- crashes/hangs/device loss;
- output instability;
- fallback warnings;
- thermal throttling or clock collapse if actually observed.

Use repeated runs and report distributions/medians per `BENCHMARKING.md`; do not promote a single best run.

Warm-up policy, run count, prompt length, generated-token count, context, batch size, threads, mmap/mlock/offload parameters, and every other material runtime control must be identical inside a same-device backend pair unless the backend makes a setting inapplicable. Record any unavoidable mismatch explicitly.

## Phase 5 — backend capability record

Produce a machine-readable research record for each physical GPU/backend combination containing at least:

- physical GPU identity independent of backend;
- memory-resource identity/capacity;
- backend kind/version;
- driver/ICD/runtime identity;
- relevant advertised backend capabilities;
- representation/model support result;
- correctness/stability result;
- materialization/memory observations;
- prefill metrics;
- decode metrics;
- evidence freshness/timestamp;
- known limitations/failures.

Shape this as **research evidence**, not as the final production resource schema.

A future planner should be able to reason from facts equivalent to:

```text
Compute Unit X
  supports execution backend Vulkan with evidence E1
  supports execution backend HIP with evidence E2
  has Memory Resource M
```

not:

```text
Compute Unit X is a Vulkan GPU
```

## Evidence retention

Create an additive retained area such as:

`docs/investigations/vulkan-v0-a/`

Retain at minimum:

- methodology frozen before canonical benchmark results;
- hardware inventories;
- before/after software-stack records if interventions occur;
- exact probe source/build/binary identities;
- model subject identity/hash;
- workload fixtures;
- raw command lines;
- raw benchmark outputs;
- correctness/repeatability outputs;
- memory/materialization observations;
- machine-readable backend capability records;
- summary comparison tables;
- manifest/hashes;
- explicit non-claims.

Large external model files do not belong in Git; retain identities/hashes only.

## Required analysis outputs

For each same-device comparison report at minimum:

```text
AMD physical GPU A
  Vulkan result
  HIP/ROCm result or NATIVE_BACKEND_UNAVAILABLE
  Vulkan/native prefill ratio
  Vulkan/native decode ratio
  startup/materialization delta
  GPU-memory delta
  persistent-host-memory delta
  correctness/stability comparison

NVIDIA physical GPU N
  Vulkan result
  CUDA result
  Vulkan/native prefill ratio
  Vulkan/native decode ratio
  startup/materialization delta
  GPU-memory delta
  persistent-host-memory delta
  correctness/stability comparison
```

Do not calculate AMD-vs-NVIDIA ratios and call them backend ratios.

## Terminal classifications

Produce exactly one V0-A terminal:

- `V0A_MATCHED_CHARACTERIZATION_COMPLETE` — required AMD Vulkan and NVIDIA Vulkan/native evidence collected; native AMD may truthfully be unavailable if that itself is demonstrated;
- `V0A_AMD_VULKAN_BLOCKED` — the installed AMD hardware/software cannot be made to execute the frozen Vulkan subject without a scope-expanding intervention;
- `V0A_CONTROL_BLOCKED` — AMD evidence is usable but the required NVIDIA Vulkan/CUDA matched control cannot be collected safely;
- `V0A_EVIDENCE_INSUFFICIENT` — collected results cannot support a trustworthy matched characterization.

A slow Vulkan result is **not** a blocked result if it is correct and measurable.

## Acceptance criteria

- [ ] Exact `inferswarm02` AMD hardware and starting driver/Vulkan state discovered and retained.
- [ ] Exact NVIDIA control hardware/backend state retained.
- [ ] Every canonical execution proves the intended physical device/backend with no silent CPU/other-GPU fallback.
- [ ] One exact model representation and workload are frozen for matched comparison.
- [ ] AMD Vulkan correctness/stability evidence exists.
- [ ] AMD HIP/ROCm is measured on the same GPU where credibly available, or its unavailability is mechanically demonstrated rather than assumed.
- [ ] NVIDIA Vulkan and CUDA are measured on the same physical GPU.
- [ ] Prefill and decode are measured separately.
- [ ] Startup/materialization, GPU memory, and host-memory behavior are recorded.
- [ ] Repeated-run distributions are retained; no best-run cherry-picking.
- [ ] Backend capability records keep physical resource identity separate from execution backend identity.
- [ ] No InferSwarm planner/runtime architecture is changed.
- [ ] Accepted #117/#133/#137 evidence is untouched.
- [ ] Results are retained with hashes/manifests and explicit non-claims.

## Completion report

Include:

1. starting InferSwarm main SHA;
2. exact nodes and physical GPU IDs/BDFs/SKUs/VRAM;
3. starting and final driver/ICD/runtime state;
4. every material host change made;
5. pinned llama.cpp commit/build/binary hashes;
6. exact model subject/hash and workload;
7. backend/device enumeration proof;
8. correctness/repeatability findings per backend;
9. same-device AMD Vulkan/native results;
10. same-device NVIDIA Vulkan/CUDA results;
11. prefill/decode/startup/memory distributions and ratios;
12. failures/fallbacks/regressions observed;
13. retained evidence paths/manifests;
14. exact V0-A terminal;
15. recommendation for V0-B;
16. confirmation no active Issue #117 correctness authority/evidence was modified.

If repository files are added, open a PR and leave it **OPEN / UNMERGED** for maintainer review. Physical characterization may complete independently of merge.
